### PR TITLE
[WIP] Added some Callback functions 

### PR DESCRIPTION
**Pull request by AdarshKumar712**
I have added the following callback functions:
1. terminateOnNaN
2. HistoryCallback
3. lrdecay
4. ModelCheckpoint

As there aren't any built-in callback functions existing in Flux, I hope these may be helpful in improving the Flux-user experience. Reference taken from keras.

### CODE REVIEW

File: src/callbacks.jl [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    x,y: could be training data (x,y) or can be different
+
+    factor: factor by which learning rate is reduced in every updation. Default set to 0.2
+    
+    loss: function to be used to calculate loss, when monitor set to 'loss'. Otherwise, optional.
+    
+    accuracy_fn: function to be used to calculate accuracy, when monitor set to 'acc'. Otherwise, optional
+    
+    patience: number of epochs that produced the monitored quantity with no improvement, 
+	      after which learning rate will be reduced or training stopped. Default value set to 5.
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: I will change that according to documentation rules.

File: src/callbacks.jl [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+Function to initialise global variables required in various callback functions.
+"""
+function init_cb()
+    global ctr = 0                     #global counter for how many times the callback has been called
+    global stop_itr = false
+    global loss_history = []
+    global best_acc = 0
+    global best_loss = Inf
+    global last_improvement = 0 
+end
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Actually there can be multiple methods that can be used for training, thus there isn't a proper way to keep counter of how many times a callback a called. So defined this ctr. However as you say above that in case of multiple callbacks, this will get affected. I am thinking to use some other way to keep the counter. Is there any way possible to make the callbacks aware of multiple callbacks?

File: src/callbacks.jl [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    
+    patience: number of epochs that produced the monitored quantity with no improvement, 
+	      after which learning rate will be reduced or training stopped. Default value set to 5.
+    
+    min_lr: lower bound on the learning rate. If no improvement seen even below that after 'patience' number of steps, the training is terminated.
+    
+    monitor: Quantity to be monitored for the provided (x,y). Can take values 'acc' or 'loss'. Default set to 'loss'
+"""
+
+function lrdecay(x,y;factor = 0.2,loss = loss,accuracy_fn = nothing,patience=5,min_lr = 1e-5,monitor="loss")
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Yes, I also think now that, there should be just metric function for that. Thanks for the suggestion. I will make the likewise changes to model_checkpoint as well

File: src/callbacks.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Would it be better to give each of the callbacks with no. of epochs requirement a seperate counter like CALLBACK_VALUES["ctr_model_check"]?

File: src/callbacks.jl [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+           @info "NaN loss, Terminating execution!!!!"
+           stop()
+           end
+       end
+
+"""
+    history(x,y,accuracy)
+
+Callback that records loss and accuracy into an array. Array can be accessed using `loss_history`.
+Arguments:
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: I will change that according to the documentation rules

File: src/callbacks.jl [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+function lrdecay(x,y;factor = 0.2,loss = loss,accuracy_fn = nothing,patience=5,min_lr = 1e-5,monitor="loss")
+            global ctr+=1
+            global best_acc,last_improvement,best_loss,stop_itr
+            if monitor == "acc"
+               acc = accuracy_fn(x,y)
+               if acc>best_acc
+                   best_acc = acc
+                   last_improvement = ctr
+               elseif (ctr-last_improvement)>patience
+                   if opt.eta>min_lr
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Oh! It should have been a input parameter. Actually while testing, I took `opt` as my Optimizer that's why it didn't throw any error that time. I will add Optimizer as a parameter

File: src/callbacks.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Should the keyword arguments be also wrapped with ` ?

File: src/callbacks.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnnychen94) [reply src/callbacks.jl]: It might also be helpful to write callbacks as function-like struct so that we can easily generate many callbacks from the template. This can achieve great flexibility when we try to compose different basic callbacks (logger, terminator, etc). The only problem is how to keep public data shared and private data not modified by others.

We may go one step further to separate callbacks into two parts:

* an internal `callback_counter` that modifies `CALLBACK_VALUES`, and let this called first at each callback pipeline.
* callback watchers(`terminateOnNaN`) that get info from `CALLBACK_VALUES` for shared values(e.g., `ctr`, `epochs`) and only update private values(e.g., `metric`).

It occurs to me that `train!` should call `_init_cb` before the real loop to clear all the previous cb status.

I'm not sure if there's better solution, the one I can come up with is to create a registry for callbacks. That can be something like:

```julia
# a very scratchy code snippet

# there won't be many callbacks, so we can make this list increment-only
const CALLBACK_REGISTRY = []

struct TerminateOnNaN
    metric
    id::UInt
end

TerminateOnNaN(metric)
    # not thread-safe IIUC but we'd probably not hit that issue since callbacks are mostly generated before we feed into `train!`
    push!(CALLBACK_REGISTRY, metric)
    TerminateOnNaN(metric, length(CALLBACK_REGISTRY))
end

function (cb::TerminateOnNaN)(x, y)
    ...
    CALLBACK_VALUES["cb_$(cb.id)"]["best_metric"] = best_metric
end
```

There're many implementation details wrt this, but here's what I would do when I try to create a framework for callbacks. How do others think? @oxinabox @CarloLucibello

**Comment by oxinabox** [comment]
The way I would do this is to have each callback hold state.
This lets us avoid using any globals,
which makes it much easier to reason about, more performance (in the non-const case),
and avoids issues e.g. with training two models at once and them both writing to same global.

For adhoc use I use closures.
```
function make_history_callback(calc_metric)
    history = [calc_metric()]  # store initial loss
    return ()->push!(history, calc_metric())
end
```
which I would use:

```
const test_x, test_y = ...
const train_data = ...
const model = Chain(...)
test_loss() = mse(test_y, model(test_x))

const history_callback = make_history_callback(test_loss)

# may have signature a bit wrong but I trust you get the point
Flux.train!(mse, Iterators.repeat(train_data, 100), model, params(model), ADAM(); callback=history_callback)

using Plots
plot(history_callback.history)
```

To be most explict than using closures would be to use functors.

```
struct HistoryCallback{F,T}
    calc_metric::F
    history::Vector{T}
end

HistoryCallback(calc_metric) = HistoryCallback(calc_metric, [calc_metric])

(cb::HistoryCallback)()= push!(cb.history, calc_metric())
```

Which i would use basically as before:
```
const history_callback = HistoryCallback(test_loss)
```

**Comment by johnnychen94** [comment]
> ```julia
> struct HistoryCallback{F,T}
>    calc_metric::F
>    history::Vector{T}
> end

Apparently it should be done like this!

**Comment by AdarshKumar712** [comment]
Thanks for the suggestions! I will make the changes accordingly.

File: src/callbacks.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (bhvieira) [reply src/callbacks.jl]: Doing it that way requires the forward-pass twice in the iteration the callback is called. I'm not sure how to merge the Zygote pullback stuff with Flux callbacks though.
BTW, I think most uses of `train!` will terminate anyways when a NaN loss emerges.

**Comment by AdarshKumar712** [comment]
@johnnychen94 I have updated the functions. Please have a look.
However, still some tasks are left.
1. How to ensure that training should stop, once `stop()` is called from the functions?
2. Documentation need improvement. Some lines are too long, need to split them.
 Can you please suggest how these tasks can be addressed?

File: src/callbacks.jl [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    terminateOnNaN(calc_metric)
+
+Callback that terminates training when `NaN` is encountered for `calc_metric` value.
+
+"""
+mutable struct terminateOnNaN{F}
+    calc_metric::F
+end
+
+function (cb::terminateOnNaN)()
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: That was one of the things that I did previously, but I observed that in general callbacks are called without any arguments and generally the data  is provided within the metric definition like 
`test_loss() = mse(test_y, model(test_x))` by @oxinabox, and even in documentation examples. However still, I guess considering your suggestion, we can provide a proper definition for `calc_metric` where we can still put suggestion for something like `calc_metric =()-> metric(valid_x, valid_y, model)` where metric can defined in your suggested manner. In that way we can provide a better intuition about calc_metric

File: src/callbacks.jl [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    lrdecay(calc_metric, opt; factor=0.2, patience=5, descend = true, min_lr=1e-5)
+
+Learning Rate Decay based on previous performance. Reduce learning rate when a metric has stopped improving.
+By default, it watches if `calc_metric` descends.
+
+# Arguments
+- `calc_metric`: the metric to be monitored. Evaluated at every call of callback function   
+- `opt`: the optimizer used while training
+- `factor::Float64=0.2`: factor by which learning rate is reduced in every updation.
+- `patience::Int=5`: number of epochs that produced the monitored 'calc_metric' with no improvement, after which learning rate will be reduced or training stopped.
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Yes, I guess that would be a better idea. I will split it that way, as I think it will be more readable that way, especially for someone who is going through the code. Otherwise I think both are parsed same way in documentation

File: src/callbacks.jl [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    best_metric::Float64
+    ctr::Int
+    factor::Float64
+    last_improvement::Int
+    descend::Bool
+    patience::Int
+    min_lr::Float64
+end
+
+lrdecay(calc_metric, opt; factor=0.2, patience=5, descend = true, min_lr=1e-6) = lrdecay(calc_metric, opt, calc_metric(), 0, factor, 0, descend, patience, min_lr)
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Oh! I haven't seen that earlier. So, should I remove this then? 
In place, I think we can have something called LRscheduler, where user can provide custom functions to schedule the LR decay, over epochs.

File: src/callbacks.jl [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+function (cb::ModelCheckpoint)()
+    cb.ctr += 1
+    metric = cb.calc_metric()
+    epoch = cb.ctr
+    if (metric<cb.best_metric && cb.descend == true) || (metric>cb.best_metric && cb.descend == false)
+        if cb.verbose==1
+            path = cb.savepath
+            filename = cb.filename
+            if cb.save_best_model_only ==1
+                @warn(" -> Monitored metric improved ! Saving model out to $path$filename")
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Yeah! that's a better idea. I will change that

File: src/callbacks.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (johnnychen94) [reply src/callbacks.jl]: For this PR I think it's better to remove this, but if you feel the LRScheduler way more friendly to end-users, open another issue for that, although I think these duplicated functionalities would be rejected.

File: src/callbacks.jl [hunk 10]
```diff
@@ -0,0 +1,146 @@
+using .Optimise: stop
+using BSON
+
+"""
+    terminateOnNaN(calc_metric)
+
+Callback that terminates training when `NaN` is encountered for `calc_metric` value.
+
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: I will add that part. I was thinking it would be better, if we provide more detail about this through examples in main documentation.

File: src/callbacks.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: Ok, I will remove that. Thanks for pointing! :)

**Comment by CarloLucibello** [comment]
I didn't look into the details of the PR, so just two general comments:

1)  callbacks are called after *each* minibatch update (if doing SGD and no full-batch updates)
2) the callback architecture in some cases implies doing twice the computation of the forward pass, which is highly inefficient

My opinion is that we shouldn't advertise the `train!` and callback approach at all, in favor of simple for loops. We shouldn't advertise bad coding practices. That said, if @johnnychen94 or other people think this kind of functionality is useful, and since some effort has gone into this PR, we should definitely have it. Just let's be mindful of 1) and 2)

File: src/callbacks.jl [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+In order to load the saved model, use:  `BSON.@load joinpath(savepath, filename) m epoch metric` (in case of `save_best_model_only=true`)
+or `BSON.@load joinpath(savepath, string('{epoch}_', filename) m epoch metric` (in case of `save_best_model_only=false`)  
+where m represent a chain of all the sub-models of the function. To access each individual model, use m.layers[i] for ith sub-model.
+
+"""
+
+mutable struct ModelCheckpoint{F,T,S,I}
+    calc_metric::F
+    model_arr::AbstractArray{T}
```
[the remark this answers is not in the document]
Comment (AdarshKumar712) [reply src/callbacks.jl]: I think as you suggested above, I guess rather doing any of the above things, its better if we simply save on the model as it is into a dict by bson, whether it's single Chain model or multiple models, and they will be retrived in the same manner as passed. So I think here there will be no need for Chain anymore. I will simply remove the Chain and AbstractArray part.

**Comment by AdarshKumar712** [comment]
> Looks nice, my comments are more about style.
> 
> There're two more concerns (not necessarily about this PR), that:
> 
>     * as @bhvieira commented, we may need to find a way to reduce unnecessary metric computation across multiple callbacks or even get forward-pass from Zygote. Currently, all metrics are computed multiple times even if they're the same.
> 
>     * Some callbacks don't need to be called so frequently, we may need a callback scheduler type to dynamically do nothing or call the callbacks, i.e., a callback that calls callbacks.
> 
> 
> For now, I think it's okay to get others involved to see if they find more things that need to be resolved. @dhairyagandhi96 @CarloLucibello

- For the first part, I think we can have a wrapper function, please tell me if that would be meaningful or not. I have something like this in my mind.

```
function callback(metric, cb_list)
    metric_val = metric()
    for each_cb in cb_list
        cblist(metric_val)
    end
end

cb_list = [...]    #list of callbacks
cb = callback(metric, cb_list)
```
- For second point, I think we can achieve that using throttle for individual functions. Like
```
cb_list = [throttle(f1, t1), throttle(f2, t2)....]
``` 
 But I don't know exactly if this way of scheduling will be efficient or not.

**Comment by AdarshKumar712** [comment]
> I didn't look into the details of the PR, so just two general comments:
> 
>     1. callbacks are called after_each_ minibatch update (if doing SGD and no full-batch updates)
For 1), using something like throttle or some more efficient way like callback scheduler, we can minimise calling callbacks.
> 
>     2. the callback architecture in some cases implies doing twice the computation of the forward pass, which is highly inefficient 
> My opinion is that we shouldn't advertise the `train!` and callback approach at all, in favor of simple for loops. We shouldn't advertise bad coding practices. That said, if @johnnychen94 or other people think this kind of functionality is useful, and since some effort has gone into this PR, we should definitely have it. Just let's be mindful of 1) and 2)

Yes, I agree that callbacks may require twice the forward pass. But I hope that forward pass may worth its extra cost, if we use it for validating our data, that is needed to be done in many models. 
Also, I think these functions aren't just meant for `train!` function but can also be equally utilised in simple custom for loops while `validating` the data. I hope they could be used more as a functionality to support easy validation and further provide checks on that validation(like saving best model).

**Comment by johnnychen94** [comment]
I suggest holding this PR a while until https://github.com/FluxML/Flux.jl/pull/1017 gets merged

**Comment by CarloLucibello** [comment]
closing as outdated
